### PR TITLE
Updates Evergreen's file handling to use NIO. This is quite a big change, so I wouldn't mind a second pair of eyes.

### PR DESCRIPTION
**Pull request by philjessies**

### CODE REVIEW

File: evergreen/src/e/edit/ExternalTools.java [hunk 1]
```diff
@@ -128,7 +144,10 @@ private static void parseFile(File file, List<ExternalToolAction> newAllTools, L
         boolean needsFile = false;
         boolean showOnPopUpMenu = false;
         
-        for (String line : StringUtilities.readLinesFromFile(file)) {
+        for (String line : Files.readAllLines(file)) {
+            if (line == "" || line.startsWith("#")) {
```
Comment (ritschwumm) [inline evergreen/src/e/edit/ExternalTools.java:148]: `line == ""` looks like a bug - i don't think comparing references is the right thing to do here.

File: evergreen/src/e/edit/WorkspaceFileList.java [hunk 2]
```diff
@@ -181,11 +259,34 @@ private void fireListeners(final boolean isNowValid) {
...
-                    public void run() {
-                        l.fileListStateChanged(isNowValid);
-                    }
-                });
+                GuiUtilities.invokeLater(() -> { l.fileListStateChanged(isNowValid); });
+            }
+        }
+    }
+    
+    private void fireFileCreated(final String filename) {
```
Comment (ritschwumm) [inline evergreen/src/e/edit/WorkspaceFileList.java:267]: this and the following 2 methods maybe deserve a helper method like this:
```
private void notifyListeners(action:Consumer<Listener>) {
     for (final Listener l : listeners) {
        GuiUtilities.invokeLater(() -> action(l));
    }
}
```
which then can be called with
```
notifyListeners(l ->  l.fileCreated(filename));
```

File: salma-hayek/src/e/util/ProcessUtilities.java [hunk 3]
```diff
@@ -93,31 +95,27 @@ public static int backQuote(final File directory, final String[] commandArray, f
...
-                        BufferedWriter out = new BufferedWriter(new OutputStreamWriter(p.getOutputStream(), "UTF-8"));
-                        out.append(input);
-                        out.flush();
-                        out.close();
-                    } catch (Exception ex) {
-                        feedExceptionToLineListener(errorLineListener, ex);
-                    }
+            Thread inputWriterThread = new Thread(() -> {
+                try {
+                    BufferedWriter out = new BufferedWriter(new OutputStreamWriter(p.getOutputStream(), "UTF-8"));
```
Comment (ritschwumm) [inline salma-hayek/src/e/util/ProcessUtilities.java:108]: should probably use try-with-resources to ensure the OutputStream is closed even when an exception is encountered

**Comment by philjessies** [comment]
Thanks - very good comments. I'm a bit embarrassed about the string
comparison one, to be honest. I blame years of C++ and Go.

Actually, I should re-base (there's been some work on head), and re-do the
pull request, as I've also made some fixes since I originally made it.

**Pull request closed without merging**